Update OSS nested tensor docs to focus on NJT (#145402)
pytorch/pytorch: docs/source/nested.rst

  The PyTorch API of nested tensors is in prototype stage and will change in the near future.

Nested tensors allow for ragged-shaped data to be contained within and operated upon as a
single tensor. Such data is stored underneath in an efficient packed representation, while exposing
a standard PyTorch tensor interface for applying operations.

A common application of nested tensors is for expressing batches of variable-length sequential data

> present in various domains (e.g. varying sentence lengths, image sizes, audio / video clip lengths).

Reviewer: ```suggestion
present in various domains (such as varying sentence lengths, image sizes, audio / video clip lengths).
```

>>> a
tensor([0, 1, 2])
>>> b
tensor([3, 4, 5, 6, 7])
>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
>>> print([component for component in nt])
[tensor([0, 1, 2]), tensor([3, 4, 5, 6, 7])]

Each tensor in the list must have the same number of dimensions, but the shapes can otherwise vary
along a single dimension. If the dimensionalities of the input components don't match, the
constructor throws an error.

>>> a = torch.randn(50, 128) # 2D tensor
>>> b = torch.randn(2, 50, 128) # 3D tensor
>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
...
RuntimeError: When constructing a nested tensor, all tensors in list must have the same dim

During construction, dtype, device, and whether gradients are required can be chosen via the
usual keyword arguments.

>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged, dtype=torch.float32, device="cuda", requires_grad=True)
>>> print([component for component in nt])
[tensor([0., 1., 2.], device='cuda:0',
       grad_fn=<UnbindBackwardAutogradNestedTensor0>), tensor([3., 4., 5., 6., 7.], device='cuda:0',
       grad_fn=<UnbindBackwardAutogradNestedTensor0>)]

> In the vein of ``torch.as_tensor``, ``torch.nested.as_nested_tensor`` can be used to preserve autograd

Reviewer 2: This is preexisting, but I'm not sure I want to mention `torch.as_tensor` here. It seems to not actually preserve autograd history in the actual cases you would want to use it, e.g. convert from numpy.

Author: removed mention of it

>>> nt.shape
torch.Size([2, j1, 6])
>>> nt.unsqueeze(-1).shape
torch.Size([2, j1, 6, 1])
>>> nt.unflatten(-1, [2, 3]).shape
torch.Size([2, j1, 2, 3])
>>> torch.cat([nt, nt], dim=2).shape
torch.Size([2, j1, 12])
>>> torch.stack([nt, nt], dim=2).shape
torch.Size([2, j1, 2, 6])
>>> nt.transpose(-1, -2).shape
torch.Size([2, 6, j1])

Attention mechanisms
^^^^^^^^^^^^^^^^^^^^

As variable-length sequences are common inputs to attention mechanisms, nested tensors support
important attention operators
`Scaled Dot Product Attention (SDPA) <https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html>`_ and
`FlexAttention <https://pytorch.org/docs/stable/nn.attention.flex_attention.html#module-torch.nn.attention.flex_attention>`_.
See
`here <https://pytorch.org/tutorials/intermediate/transformer_building_blocks.html#multiheadattention>`__
for usage examples of NJT with SDPA and
`here <https://pytorch.org/tutorials/intermediate/transformer_building_blocks.html#flexattention-njt>`__
for usage examples of NJT with FlexAttention.

.. _usage_with_torch_compile:

Usage with torch.compile
++++++++++++++++++++++++

> NJTs are designed to be used with ``torch.compile()`` for optimal performance. In particular,

Reviewer 2: "NJTs are designed to be used with ``torch.compile()`` for optimal performance."

Could another way of wording this be something like:
- torch.compile can be used to speed up programs using NJTs. This can be especially important for programs involving smaller tensor sizes where the subclass overhead of NJT adds significant part to runtime.

I think it would be less scary for the people who can't use compile, but may otherwise still benefit.

Author: Part of my reasoning behind making a very bold statement is to shape expectations for performance with NJTs. It's been the case in the past that people use NJTs in eager expecting speedups but fall into the overhead-bound regime and are surprised to find that the result it slower than whatever they were doing before. The idea here was to be very explicit that if you're not using torch.compile, you shouldn't expect NJT to help you.

But I'm also okay with being more nuanced here. The downside is that the dependence on torch.compile is now multi-faceted and it's tricky to make accurate statements about when the user can expect speedups in eager. Consider the padded dense conversion logic utilized behind binary broadcasting and some other ops - in eager, we don't have fusion of the "computation sandwich" and so overhead and memory usage would on par with or worse than manual padded computation in eager.

``torch.compile()`` is important to eliminate eager-mode python tensor subclass overhead, which

> affects NJT as it is a python tensor subclass. NJTs work out-of-the-box and

Reviewer 2: I wonder if it is worth mentioning this explicitly, since maybe that is the expectation already when working with compile.

Author: true, it should be the expectation. I'm making it explicit to be clear about what users should expect from NJT. There are other features within PyTorch that they can't expect to compose with torch.compile without issue, so it's nice to set this one apart via "you should expect a seamless experience with torch.compile"

graph-break-free both when passed as inputs to a compiled function or module OR when
instantiated in-line within the function.

>>> import torch
>>> a = torch.randn(2, 3)
>>> b = torch.randn(4, 3)
>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
>>> def f(x): return x.sin() + 1
...
>>> compiled_f = torch.compile(f, fullgraph=True)
>>> output = compiled_f(nt)
>>> output.shape
torch.Size([2, j1, 3])
>>> def g(values, offsets): return torch.nested.nested_tensor_from_jagged(values, offsets) * 2.
...
>>> compiled_g = torch.compile(g, fullgraph=True)
>>> output2 = compiled_g(nt.values(), nt.offsets())
>>> output2.shape
torch.Size([2, j1, 3])

> By default, NJTs support

Reviewer 2: Remove "By default"? (not sure we can disable it)

>>> nt2 = torch.nested.nested_tensor_from_jagged(values=torch.randn(82, 128), offsets=nt1.offsets())
>>> nt3 = nt1 + nt2
>>> nt3.shape
torch.Size([2, j1, 128])

Data dependent operatation within torch.compile
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

    torch._dynamo.exc.Unsupported: data dependent operator: aten._local_scalar_dense.default; to enable, set torch._dynamo.config.capture_scalar_outputs = True

This error occurs when calling an op that does data-dependent operation within torch.compile; this
commonly occurs for ops that need to examine the values of the NJT's ``offsets`` to determine the
output shape. For example:

>>> a = torch.randn(50, 128)
>>> b = torch.randn(32, 128)
>>> nt = torch.nested.nested_tensor([a, b], layout=torch.jagged, dtype=torch.float32)
>>> def f(nt): return nt.chunk(2, dim=0)[0]
...
>>> compiled_f = torch.compile(f, fullgraph=True)
>>> output = compiled_f(nt)

In this example, calling ``chunk()`` on the batch dimension of the NJT requires examination of the
NJT's ``offsets`` data to delineate batch item boundaries within the packed ragged dimension. The
workaround here is as suggested in the error message.

> >>> torch._dynamo.config.capture_scalar_outputs = True

Reviewer 2: Should we also mention `capture_dynamic_output_shape_ops=True` here?

Author: yeah good point, probably should! added
